File: lib-database/src/main/java/com/ustadmobile/core/db/dao/CustomFieldValueOptionDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.CustomFieldValueOption;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.util.List;

@UmDao(insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN,
        updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class CustomFieldValueOptionDao
    implements SyncableDao<CustomFieldValueOption, CustomFieldValueOptionDao> {

    @UmQuery("SELECT * FROM CustomFieldValueOption " +
            " WHERE customFieldValueOptionFieldUid = :customFieldUid " +
            " AND customFieldValueOptionActive = 1")
    public abstract UmProvider<CustomFieldValueOption> findAllOptionsForField(long customFieldUid);

    @UmQuery("SELECT * FROM CustomFieldValueOption " +
            " WHERE customFieldValueOptionFieldUid = :customFieldUid " +
            " AND customFieldValueOptionActive = 1")
    public abstract void findAllOptionsForFieldAsync(long customFieldUid, UmCallback<List<CustomFieldValueOption>> optionsCallback);

    @UmQuery("UPDATE CustomFieldValueOption SET customFieldValueOptionActive = 0 WHERE" +
            " customFieldValueOptionUid = :uid")
    public abstract void deleteOption(long uid, UmCallback<Integer> resultObjecT);


    @UmUpdate
    public abstract void updateAsync(CustomFieldValueOption entity, UmCallback<Integer> resultObject);

    @UmQuery("SELECT * FROM CustomFieldValueOption WHERE customFieldValueOptionUid = :uid")
    public abstract void findByUidAsync(long uid, UmCallback<CustomFieldValueOption> resultObject) ;
}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/CustomFieldValueOptionDao.kt
package com.ustadmobile.core.db.dao

import androidx.paging.DataSource
import androidx.room.Query
import androidx.room.Update
import com.ustadmobile.core.impl.UmCallback
import com.ustadmobile.lib.database.annotation.UmDao
import com.ustadmobile.lib.database.annotation.UmRepository
import com.ustadmobile.lib.db.entities.CustomFieldValueOption

@UmDao(insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN, 
        updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
abstract class CustomFieldValueOptionDao : BaseDao<CustomFieldValueOption> {

    @Query("SELECT * FROM CustomFieldValueOption " +
            " WHERE customFieldValueOptionFieldUid = :customFieldUid " +
            " AND customFieldValueOptionActive = 1")
    abstract fun findAllOptionsForField(customFieldUid: Long): DataSource.Factory<Int, CustomFieldValueOption>

    @Query("SELECT * FROM CustomFieldValueOption " +
            " WHERE customFieldValueOptionFieldUid = :customFieldUid " +
            " AND customFieldValueOptionActive = 1")
    abstract fun findAllOptionsForFieldAsync(customFieldUid: Long, optionsCallback: UmCallback<List<CustomFieldValueOption>>)

    @Query("UPDATE CustomFieldValueOption SET customFieldValueOptionActive = 0 WHERE" + 
            " customFieldValueOptionUid = :uid")
    abstract fun deleteOption(uid: Long, resultObjecT: UmCallback<Int>)


    @Update
    abstract fun updateAsync(entity: CustomFieldValueOption, resultObject: UmCallback<Int>)

    @Query("SELECT * FROM CustomFieldValueOption WHERE customFieldValueOptionUid = :uid")
    abstract fun findByUidAsync(uid: Long, resultObject: UmCallback<CustomFieldValueOption>)
}
